Mac ARM CPU support fails looking for nvcc

With #7 (CPU fallback) implemented I thought I would try it out.

When I run `pip install .` it complains about not finding `nvcc`, which is unexpected:

Environment creation:
```bash
conda create -n protax-gpu pip pybind11 pandas
conda activate protax-gpu
pip install --upgrade "jax[cpu]" chex
```

Wheel creation:

```bash
cd ~/src/PROTAX-GPU
pip install .
Processing /Users/gwtaylor/src/PROTAX-GPU
  Preparing metadata (setup.py) ... done
Requirement already satisfied: jax in /opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages (from protax==0.0.0) (0.4.20)
Requirement already satisfied: jaxlib in /opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages (from protax==0.0.0) (0.4.20)
Requirement already satisfied: ml-dtypes>=0.2.0 in /opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages (from jax->protax==0.0.0) (0.3.1)
Requirement already satisfied: numpy>=1.22 in /opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages (from jax->protax==0.0.0) (1.26.0)
Requirement already satisfied: opt-einsum in /opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages (from jax->protax==0.0.0) (3.3.0)
Requirement already satisfied: scipy>=1.9 in /opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages (from jax->protax==0.0.0) (1.11.4)
Building wheels for collected packages: protax
  Building wheel for protax (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [80 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.macosx-11.0-arm64-cpython-312
      creating build/lib.macosx-11.0-arm64-cpython-312/protax
      copying src/protax/protax_utils.py -> build/lib.macosx-11.0-arm64-cpython-312/protax
      copying src/protax/baseline.py -> build/lib.macosx-11.0-arm64-cpython-312/protax
      copying src/protax/classify.py -> build/lib.macosx-11.0-arm64-cpython-312/protax
      copying src/protax/__init__.py -> build/lib.macosx-11.0-arm64-cpython-312/protax
      copying src/protax/model.py -> build/lib.macosx-11.0-arm64-cpython-312/protax
      copying src/protax/taxonomy.py -> build/lib.macosx-11.0-arm64-cpython-312/protax
      creating build/lib.macosx-11.0-arm64-cpython-312/knn_jax
      copying src/knn_jax/knn_jax.py -> build/lib.macosx-11.0-arm64-cpython-312/knn_jax
      copying src/knn_jax/__init__.py -> build/lib.macosx-11.0-arm64-cpython-312/knn_jax
      running egg_info
      creating protax.egg-info
      writing protax.egg-info/PKG-INFO
      writing dependency_links to protax.egg-info/dependency_links.txt
      writing requirements to protax.egg-info/requires.txt
      writing top-level names to protax.egg-info/top_level.txt
      writing manifest file 'protax.egg-info/SOURCES.txt'
      reading manifest file 'protax.egg-info/SOURCES.txt'
      writing manifest file 'protax.egg-info/SOURCES.txt'
      running build_ext
      -- The CXX compiler identification is AppleClang 15.0.0.15000040
      CMake Error at /opt/homebrew/Cellar/cmake/3.27.7/share/cmake/Modules/CMakeDetermineCUDACompiler.cmake:180 (message):
        Failed to find nvcc.

        Compiler requires the CUDA toolkit.  Please set the CUDAToolkit_ROOT
        variable.
      Call Stack (most recent call first):
        CMakeLists.txt:2 (project)


      -- Configuring incomplete, errors occurred!
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/Users/gwtaylor/src/PROTAX-GPU/setup.py", line 100, in <module>
          setup(
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/__init__.py", line 103, in setup
          return distutils.core.setup(**attrs)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 185, in setup
          return run_commands(dist)
                 ^^^^^^^^^^^^^^^^^^
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
          dist.run_commands()
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 969, in run_commands
          self.run_command(cmd)
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/dist.py", line 989, in run_command
          super().run_command(command)
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/wheel/bdist_wheel.py", line 369, in run
          self.run_command("build")
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
          self.distribution.run_command(command)
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/dist.py", line 989, in run_command
          super().run_command(command)
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/command/build.py", line 131, in run
          self.run_command(cmd_name)
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/cmd.py", line 318, in run_command
          self.distribution.run_command(command)
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/dist.py", line 989, in run_command
          super().run_command(command)
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/dist.py", line 988, in run_command
          cmd_obj.run()
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/command/build_ext.py", line 88, in run
          _build_ext.run(self)
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/setuptools/_distutils/command/build_ext.py", line 345, in run
          self.build_extensions()
        File "/Users/gwtaylor/src/PROTAX-GPU/setup.py", line 56, in build_extensions
          subprocess.check_call(
        File "/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/subprocess.py", line 413, in check_call
          raise CalledProcessError(retcode, cmd)
      subprocess.CalledProcessError: Command '['cmake', '/Users/gwtaylor/src/PROTAX-GPU', '-DCMAKE_INSTALL_PREFIX=/Users/gwtaylor/src/PROTAX-GPU/build/lib.macosx-11.0-arm64-cpython-312', '-DPython_EXECUTABLE=/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/bin/python3.12', '-DPython_LIBRARIES=/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/libpython3.12.a', '-DPython_INCLUDE_DIRS=/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/include/python3.12', '-DCMAKE_BUILD_TYPE=Release', '-DCMAKE_PREFIX_PATH=/opt/homebrew/Caskroom/miniforge/base/envs/protax-gpu/lib/python3.12/site-packages/pybind11/share/cmake/pybind11']' returned non-zero exit status 1.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for protax
  Running setup.py clean for protax
Failed to build protax
ERROR: Could not build wheels for protax, which is required to install pyproject.toml-based projects
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,5 +1,5 @@
 cmake_minimum_required(VERSION 3.10)
-project(knn_jax LANGUAGES CXX CUDA)
+project(knn_jax LANGUAGES CXX)
 
 message(STATUS "Using CMake version " ${CMAKE_VERSION})
 message(${CMAKE_SOURCE_DIR})
@@ -8,19 +8,36 @@ find_package(Python COMPONENTS Interpreter Development REQUIRED)
 find_package(pybind11 CONFIG REQUIRED)
 
 include_directories(${CMAKE_CURRENT_LIST_DIR}/lib)
-include_directories(${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES})
 
-# cpu op library
+# if (CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64" AND CMAKE_SYSTEM_NAME MATCHES "Darwin")
+if (CMAKE_SYSTEM_NAME MATCHES "Darwin")
+    message(STATUS "Mac detected. Skipping CUDA configuration.")
+    set(CUDA_FOUND OFF)
+else()
+    enable_language(CUDA)
+    find_package(CUDA)
+    if (CUDA_FOUND)
+        include_directories(${CMAKE_CUDA_TOOLKIT_INCLUDE_DIRECTORIES})
+    else()
+        message(WARNING "CUDA not found or not supported. Skipping GPU ops.")
+    endif()
+endif()
+
+# CPU op library
 pybind11_add_module(
   cpu_ops
   ${CMAKE_CURRENT_LIST_DIR}/lib/knn_cpu.cpp
 )
 install(TARGETS cpu_ops DESTINATION knn_jax)
 
-# gpu op library
-pybind11_add_module(
-  gpu_ops
-  ${CMAKE_CURRENT_LIST_DIR}/lib/knn_kernels.cu
-  ${CMAKE_CURRENT_LIST_DIR}/lib/knn_dispatch.cpp
-)
-install(TARGETS gpu_ops DESTINATION knn_jax)
+# GPU op library (only if CUDA is found)
+if(CUDA_FOUND)
+  pybind11_add_module(
+    gpu_ops
+    ${CMAKE_CURRENT_LIST_DIR}/lib/knn_kernels.cu
+    ${CMAKE_CURRENT_LIST_DIR}/lib/knn_dispatch.cpp
+  )
+  install(TARGETS gpu_ops DESTINATION knn_jax)
+else()
+  message(STATUS "Skipping GPU ops build since CUDA is not available.")
+endif()

--- a/README.md
+++ b/README.md
@@ -49,7 +49,7 @@ src/
 | --- | --- | --- | --- |
 | **Linux** | yes | yes | n/a |
 | **Mac X86_64** | yes | n/a | no |
-| **MAC (ARM)** | experimental | n/a | no |
+| **MAC (ARM)** | yes | n/a | no |
 | **Windows** | experimental | experimental | n/a |
 
 # Installation:
@@ -88,7 +88,7 @@ conda install -c conda-forge cmake
 | --- | --- | --- |
 | **Linux** | CPU | `pip install "jax[cpu]"`|
 | **Linux** | GPU | `pip install "jax[cuda122]"`
-| **MacOS** | CPU | `pip install "jax[cpu]" "jaxlib[cpu]"`|
+| **MacOS** | CPU | `pip uninstall jax jaxlib && conda install -c conda-forge jax=0.4.26 jaxlib=0.4.23`|
 | **MacOS** | GPU | **Not yet supported**|
 
 **NOTE:** The GPU installation command assumes you have already installed CUDA 12.2 as per step 1.
@@ -117,14 +117,58 @@ This will install a package called `protax` in your environment.
 
 **TODO**: Add Windows support instructions and check macOS GPU support
 
+If you are on MacOS and facing installation issues, run the following commands
+```
+chmod +x ./scripts/fix_librhash.sh
+./scripts/fix_librhash.sh
+```
+
+
 # Usage
 Instructions for running PROTAX-GPU for inference and training.
 
 ## Inference
-**TODO**
+Once you have a trained model, you can use the classify_file function to classify query sequences.
+
+Run the sequence classification script:
+```
+python scripts/process_seqs.py [PATH_TO_QUERY_SEQUENCES] [PATH_TO_MODEL] [PATH_TO_TAXONOMY]
+```
+Example:
+
+```
+python scripts/process_seqs.py data/refs.aln models/params/model.npz models/ref_db/taxonomy37k.npz
+```
+<!-- python scripts/process_seqs.py FinPROTAX/FinPROTAX/modelCOIfull/refs.aln models/params/model.npz models/ref_db/taxonomy37k.npz -->
+
+Arguments:
+
+- `PATH_TO_QUERY_SEQUENCES`: File containing the sequences to classify (e.g., FASTA or alignment file)(Can use refs.aln from [FinPROTAX](https://github.com/psomervuo/FinPROTAX/tree/main) for experiment)
+- `PATH_TO_MODEL`: Path to the model. (Base Model is available in `models/params/model.npz`)
+- `PATH_TO_TAXONOMY`: Path to the taxonomy .npz file. (taxonomy file is available in `models/ref_db/taxonomy37k.npz`)
+
+
+Results are saved to `pyprotax_results.csv`
 
 ## Training
-**TODO**
+Run the script from the command line: You need to specify the paths to your training data and target data using the `--train_dir` and `--targ_dir` arguments, respectively.
+```
+python scripts/train_model.py --train_dir [PATH_TO_TRAINING_DATA] --targ_dir [PATH_TO_TARGET_DATA]
+```
+
+Command Line Arguments
+- `--train_dir`: Path to the training data (e.g., refs.aln).
+- `--targ_dir`: Path to the target data (e.g., targets.csv).
+
+Training Configuration
+The script uses the following training parameters:
+
+- Learning rate: `0.001`
+- Batch size: `500`
+- Number of epochs: `30`
+These parameters are predefined in the script and can be modified if needed by editing the dictionary `tc` in the code.
+
+The script uses `models/params/model.npz` as baseline and saves the trained model at `models/params/m2.npz`
 
 # Hardware
 To reproduce the BOLD dataset experiments, PROTAX-GPU requires an NVIDIA GPU with at least 16GB VRAM and CUDA compute capability 6.0 or later. This corresponds to GPUs in the NVIDIA Pascal, NVIDIA Volta™, NVIDIA Turing™, NVIDIA Ampere architecture, and NVIDIA Hopper™ architecture families.

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,5 +1,5 @@
 numpy
-chex==0.1.82
+chex==0.1.85
 pybind11
 scipy
 pandas

--- a/scripts/fix_librhash.sh
+++ b/scripts/fix_librhash.sh
@@ -1,0 +1,21 @@
+#!/bin/bash
+
+# This script will find the librhash.1.dylib file and automatically create a symlink named librhash.0.dylib in the same directory. It will also confirm the symlink has been created by listing the contents of the library directory.
+# Find the path of the librhash*.dylib file
+lib_path=$(find $CONDA_PREFIX/lib/ -name "librhash*.dylib" | grep librhash.1.dylib)
+
+if [ -z "$lib_path" ]; then
+    echo "librhash.1.dylib not found."
+    exit 1
+fi
+
+# Set the path for the new symlink
+symlink_path="$CONDA_PREFIX/lib/librhash.0.dylib"
+
+# Create the symbolic link
+ln -sf "$lib_path" "$symlink_path"
+
+# Verify the symbolic link was created
+ls -l $CONDA_PREFIX/lib | grep librhash
+
+echo "Symbolic link created for librhash.0.dylib"

--- a/scripts/process_seqs.py
+++ b/scripts/process_seqs.py
@@ -1,13 +1,16 @@
 from protax.classify import classify_file
+import sys
 
 if __name__ == "__main__":
+    protax_args = sys.argv
+    if len(protax_args) < 4:
+        print(
+            "Usage: python3 classify.py [PATH_TO_TAXONOMY_FILE] [PATH_TO_PARAMETERS] [PATH_TO_QUERY_SEQUENCES]"
+        )
 
-    # protax_args = sys.argv
-    # if len(protax_args) < 4:
-    #     print("Usage: python3 classify.py [PATH_TO_TAXONOMY_FILE] [PATH_TO_PARAMETERS] [PATH_TO_QUERY_SEQUENCES]")
-    
-    # tax_dir, model_dir, query_dir = protax_args[1:4]
+    query_dir, model_dir, tax_dir = protax_args[1:4]
+    classify_file(query_dir, model_dir, tax_dir)
 
-    # testing for now
-    query_dir = r"/home/roy/Documents/PROTAX-dsets/30k_small/refs.aln"
-    classify_file(query_dir, "models/params/model.npz", "models/ref_db/taxonomy37k.npz") 
+    # testing
+    # query_dir = r"/home/roy/Documents/PROTAX-dsets/30k_small/refs.aln"
+    # classify_file(query_dir, "models/params/model.npz", "models/ref_db/taxonomy37k.npz")

--- a/scripts/train_gd.py
+++ b/scripts/train_gd.py
@@ -15,20 +15,16 @@ from pathlib import Path
 import random
 import matplotlib.pyplot as plt
 from functools import partial
-
-
-debug_f = r"/home/roy/Documents/PROTAX-dsets/30k_small/debug.aln"
-train_dir = r"/home/roy/Documents/PROTAX-dsets/30k_small/refs.aln"
-targ_dir = "/home/roy/Documents/PROTAX-dsets/30k_small/30k-targets.csv"
+import argparse
 
 
 def CE_loss(log_probs, y_ind):
     """
-    Computes the cross-entropy loss between the log_probs and 
+    Computes the cross-entropy loss between the log_probs and
     labels y_ind.
 
     Args:
-        log_probs: Log probabilities returned by the model, shape (N, D) 
+        log_probs: Log probabilities returned by the model, shape (N, D)
         y_ind: Integer array of true class indices, shape (N,)
     """
     return -jnp.sum(jnp.take(log_probs, y_ind, axis=0))
@@ -80,11 +76,13 @@ def mask_n2s(n2s, node_state, i):
     # update empty but known entries
     node_state = np.logical_or(node_state, empty)
     node_state = np.concatenate((node_state, has_refs), axis=1)
-    
-    n2s = CSRWrapper(data=jnp.array(n2s.data),
-                        indices=jnp.array(n2s.indices), 
-                        indptr=jnp.array(n2s.indptr),
-                        shape=n2s.shape)
+
+    n2s = CSRWrapper(
+        data=jnp.array(n2s.data),
+        indices=jnp.array(n2s.indices),
+        indptr=jnp.array(n2s.indptr),
+        shape=n2s.shape,
+    )
 
     return n2s, jnp.array(node_state)
 
@@ -96,9 +94,9 @@ def load_params(pdir, tdir):
     tax = np.load(tax_dir.resolve())
     par = np.load(par_dir.resolve())
 
-    beta = par['beta']
-    sc = par['scalings']
-    lvl = tax['node_layer']
+    beta = par["beta"]
+    sc = par["scalings"]
+    lvl = tax["node_layer"]
 
     tax = np.load(tax_dir.resolve())
     par = np.load(par_dir.resolve())
@@ -106,19 +104,24 @@ def load_params(pdir, tdir):
     return beta, lvl, sc
 
 
-def train(train_config):
-    tree, params, N, segnum = protax_utils.read_model_jax("models/params/model.npz", "models/ref_db/taxonomy.npz")
+def train(train_config, train_dir, targ_dir):
+    tree, params, N, segnum = protax_utils.read_model_jax(
+        "models/params/model.npz", "models/ref_db/taxonomy37k.npz"
+    )
     pkey = jax.random.PRNGKey(0)
     lr = train_config["learning_rate"]
 
     beta = jax.random.uniform(pkey, (7, 4))
-    n2s = sp.csr_matrix((tree.node2seq.data, tree.node2seq.indices, tree.node2seq.indptr), shape=tree.node2seq.shape)
+    n2s = sp.csr_matrix(
+        (tree.node2seq.data, tree.node2seq.indices, tree.node2seq.indptr),
+        shape=tree.node2seq.shape,
+    )
     targ = get_targ(targ_dir)
     seq_list, ok_list = protax_utils.read_refs(train_dir)
     node_state = np.expand_dims(np.array(tree.node_state)[:, 0], 1)
 
     # params and node lvl
-    _, lvl, sc = load_params("models/params/model.npz", "models/ref_db/taxonomy.npz")
+    _, lvl, sc = load_params("models/params/model.npz", "models/ref_db/taxonomy37k.npz")
     loss_hist = []
     for e in range(train_config["num_epochs"]):
         print(f"epoch {e}")
@@ -140,11 +143,33 @@ def train(train_config):
 
             # masks out a node2seq column given reference index (~1.2 ms)
             tree.node2seq, tree.node_state = mask_n2s(n2s, node_state, i)
-            beta_grad += f_grad(q, ok, tree, beta, params.sc_mean, params.sc_var, N, segnum, targ.at[i].get(), lvl)
-            batch_loss += forward_jit(q, ok, tree, beta, params.sc_mean, params.sc_var, N, segnum, targ.at[i].get(), lvl)
+            beta_grad += f_grad(
+                q,
+                ok,
+                tree,
+                beta,
+                params.sc_mean,
+                params.sc_var,
+                N,
+                segnum,
+                targ.at[i].get(),
+                lvl,
+            )
+            batch_loss += forward_jit(
+                q,
+                ok,
+                tree,
+                beta,
+                params.sc_mean,
+                params.sc_var,
+                N,
+                segnum,
+                targ.at[i].get(),
+                lvl,
+            )
 
             if i % train_config["batch_size"] == 0:
-                beta = beta - lr*beta_grad
+                beta = beta - lr * beta_grad
                 loss_sum += batch_loss
                 curr_loss = batch_loss / train_config["batch_size"]
                 print("batch_loss: ", curr_loss)
@@ -152,26 +177,28 @@ def train(train_config):
                 batch_loss = 0
 
                 # grad norm
-                bflat = beta_grad.reshape(beta_grad.shape[0]*beta_grad.shape[1])
-                
-        
-        
+                bflat = beta_grad.reshape(beta_grad.shape[0] * beta_grad.shape[1])
+
         print("loss: ", loss_sum / seq_list.shape[0])
-        
+
         # save checkpoint
         mf = Path("models/params/m2.npz")
-        np.savez_compressed(
-            mf.resolve(),
-            beta = np.array(beta),
-            scalings = sc
-        )
+        np.savez_compressed(mf.resolve(), beta=np.array(beta), scalings=sc)
     plt.plot(loss_hist)
     plt.show()
-    
+
 
 if __name__ == "__main__":
+    # parse config from command line
+    parser = argparse.ArgumentParser(description="Train a model")
+    parser.add_argument("--train_dir", type=str, help="Path to training data")
+    parser.add_argument("--targ_dir", type=str, help="Path to target data")
+    args = parser.parse_args()
+    train_dir = Path(args.train_dir)
+    targ_dir = Path(args.targ_dir)
 
-    # TODO parse config from command line
+    # train_dir = r"/home/roy/Documents/PROTAX-dsets/30k_small/refs.aln"
+    # targ_dir = "/home/roy/Documents/PROTAX-dsets/30k_small/30k-targets.csv"
 
     # training config
     tc = {
@@ -180,4 +207,4 @@ if __name__ == "__main__":
         "num_epochs": 30,
     }
 
-    train(tc)
+    train(tc, train_dir, targ_dir)  # train the model

--- a/setup.py
+++ b/setup.py
@@ -1,11 +1,14 @@
-import os, subprocess
+import os
+import subprocess
 
 import pybind11
 from setuptools import setup, find_packages, Extension
 from setuptools.command.build_ext import build_ext
+import platform
 
 HERE = os.path.dirname(os.path.realpath(__file__))
 pb11_dir = pybind11.get_include()
+
 
 class CMakeBuildExt(build_ext):
     def build_extensions(self):
@@ -13,13 +16,23 @@ class CMakeBuildExt(build_ext):
         Build all extensions with CMake.
         """
         # First: configure CMake build
-        import platform
         import sys
         import sysconfig
 
-
         # Work out the relevant Python paths to pass to CMake, adapted from the
         # PyTorch build system
+        install_dir = os.path.abspath(os.path.dirname(self.get_ext_fullpath("dummy")))
+        os.makedirs(install_dir, exist_ok=True)
+
+        # Determine if we're on an ARM-based Mac to disable CUDA
+        cmake_args = [
+            "-DCMAKE_INSTALL_PREFIX={}".format(install_dir),
+            "-DPython_EXECUTABLE={}".format(sys.executable),
+            "-DCMAKE_BUILD_TYPE={}".format("Debug" if self.debug else "Release"),
+            "-DCMAKE_PREFIX_PATH={}".format(pybind11.get_cmake_dir()),
+        ]
+
+        # Specific handling for Windows platform
         if platform.system() == "Windows":
             cmake_python_library = "{}/libs/python{}.lib".format(
                 sysconfig.get_config_var("prefix"),
@@ -30,40 +43,32 @@ class CMakeBuildExt(build_ext):
                     sys.base_prefix,
                     sysconfig.get_config_var("VERSION"),
                 )
+            cmake_args.append("-DPython_LIBRARIES={}".format(cmake_python_library))
         else:
             cmake_python_library = "{}/{}".format(
                 sysconfig.get_config_var("LIBDIR"),
                 sysconfig.get_config_var("INSTSONAME"),
             )
-        cmake_python_include_dir = sysconfig.get_paths()['include']
-
-        install_dir = os.path.abspath(
-            os.path.dirname(self.get_ext_fullpath("dummy"))
+        cmake_python_include_dir = sysconfig.get_paths()["include"]
+        cmake_args.extend(
+            [
+                "-DPython_LIBRARIES={}".format(cmake_python_library),
+                "-DPython_INCLUDE_DIRS={}".format(cmake_python_include_dir),
+            ]
         )
-        os.makedirs(install_dir, exist_ok=True)
-        cmake_args = [
-            "-DCMAKE_INSTALL_PREFIX={}".format(install_dir),
-            "-DPython_EXECUTABLE={}".format(sys.executable),
-            "-DPython_LIBRARIES={}".format(cmake_python_library),
-            "-DPython_INCLUDE_DIRS={}".format(cmake_python_include_dir),
-            "-DCMAKE_BUILD_TYPE={}".format(
-                "Debug" if self.debug else "Release"
-            ),
-            "-DCMAKE_PREFIX_PATH={}".format(pybind11.get_cmake_dir()),
-        ]
+
+        # Add condition for ARM-based Macs to disable CUDA
+        if platform.system() == "Darwin" and platform.machine() == "x86_64":
+            cmake_args.append("-DUSE_CUDA=OFF")
 
         os.makedirs(self.build_temp, exist_ok=True)
-        subprocess.check_call(
-            ["cmake", HERE] + cmake_args, cwd=self.build_temp
-        )
+        subprocess.check_call(["cmake", HERE] + cmake_args, cwd=self.build_temp)
 
-        # Build all the extensions
+        # Call CMake build for all extensions
         super().build_extensions()
 
-        # Finally run install
         subprocess.check_call(
-            ["cmake", "--build", ".", "--target", "install"],
-            cwd=self.build_temp,
+            ["cmake", "--build", ".", "--target", "install"], cwd=self.build_temp
         )
 
     def build_extension(self, ext):
@@ -76,39 +81,40 @@ class CMakeBuildExt(build_ext):
             cwd=self.build_temp,
         )
 
-# TODO: check if gpu support is available
 
 extensions = [
     Extension(
         name="knn_jax.cpu_ops",
-        sources=["lib/knn_cpu.cpp"]
-    ),
-    Extension(
-        name="knn_jax.gpu_ops",
-        sources=['lib/knn_dispatch.cpp', 'lib/knn_kernels.cu'],
+        sources=["lib/knn_cpu.cpp"],
         include_dirs=[pb11_dir],
     ),
 ]
 
-pdirs = {
-    "protax": "src/protax",
-    "knn_jax": "src/knn_jax"
-}
+# Add GPU ops only if not on ARM-based Mac
+if platform.system() != "Darwin":
+    extensions.append(
+        Extension(
+            name="knn_jax.gpu_ops",
+            sources=["lib/knn_dispatch.cpp", "lib/knn_kernels.cu"],
+            include_dirs=[pb11_dir],
+        )
+    )
+
+pdirs = {"protax": "src/protax", "knn_jax": "src/knn_jax"}
 
 
 # https://setuptools.pypa.io/en/latest/references/keywords.html#keywords
 setup(
-    name="protax",                                  # name of package
+    name="protax",  # name of package
     author="Roy Li",
     author_email="roymy.li12@gmail.com",
     url="github.com/uoguelph-mlrg/protax-gpu",
     # license="MIT",
-    description=(
-        "GPU accelerated DNA barcoding"
-    ),
+    description=("GPU accelerated DNA barcoding"),
     long_description_content_type="text/markdown",
-    packages=["protax", "knn_jax"],                    # packages to include in final knn_jax package
-    package_dir=pdirs,                          # root directory for included python packages
+    # packages to include in final knn_jax package
+    packages=["protax", "knn_jax"],
+    package_dir=pdirs,  # root directory for included python packages
     include_package_data=True,
     install_requires=["jax", "jaxlib"],
     extras_require={"test": "pytest"},

--- a/src/knn_jax/knn_jax.py
+++ b/src/knn_jax/knn_jax.py
@@ -2,20 +2,29 @@
 # This file defines the interface between JAX and the
 # custom KNN functions that are implemented in C++/CUDA.
 # =======================================================
-
 __all__ = ["knn, knn_v2"]
+import sys
+import site
 
-from . import gpu_ops, cpu_ops                        # refers to knn_jax installed in site-packages
+# Get site-packages path and add to sys.path
+site_packages = site.getsitepackages()[0]  # This will point to site-packages
+sys.path.append(site_packages)
+sys.path.append(f"{site_packages}/knn_jax")
+import cpu_ops
+
 from jax.lib import xla_client
 from jax import core
 from jax.core import ShapedArray
 from jax.interpreters import xla, mlir, batching
-from jaxlib.hlo_helpers import custom_call            # NOTE: change to mhlo_helpers in older versions of JAX
+from jaxlib.hlo_helpers import (
+    custom_call,
+)  # NOTE: change to mhlo_helpers in older versions of JAX
 from functools import partial
 
 import numpy as np
 import jax.numpy as jnp
 from jax.experimental import sparse
+
 
 # =======================================================
 #                      Helper Functions
@@ -26,9 +35,11 @@ def default_layouts(*shapes):
     """
     return [range(len(shape) - 1, -1, -1) for shape in shapes]
 
+
 # =======================================================
 #                     Primitives
 # =======================================================
+
 
 # Expose primitive to user code
 def knn(indptr, indices, matdat, N):
@@ -83,23 +94,21 @@ def _knn_lowering(ctx, indptr, indices, matdat, res, platform="cpu"):
     # dispatch for float32 only
     if mat_dtype != jnp.float32:
         raise NotImplementedError(f"unsupported dtype: {mat_dtype}")
-
-    if gpu_ops is None:
-        raise ValueError("gpu_ops not compiled")
-
     if platform == "gpu":
+        import gpu_ops
+
         # create opaque descriptor for problem size
         opaque = gpu_ops.build_knn_descriptor(N)
         out = custom_call(
-            b"gpu_knn_f32",                                   # call target name
-            out_types=[res_type],
-            operands=[indptr, indices, matdat], 
-            operand_layouts=default_layouts(ip_type.shape,
-                                            idx_type.shape, 
-                                            md_type.shape),
-            result_layouts=default_layouts(res_type.shape),   # memory layout
-            backend_config=opaque,                            # opaque descriptor
-        )
+            b"gpu_knn_f32",  # call target name
+            result_types=[res_type],
+            operands=[indptr, indices, matdat],
+            operand_layouts=default_layouts(
+                ip_type.shape, idx_type.shape, md_type.shape
+            ),
+            result_layouts=default_layouts(res_type.shape),  # memory layout
+            backend_config=opaque,  # opaque descriptor
+        ).results
         # output must be iterable
         return [out]
 
@@ -107,16 +116,14 @@ def _knn_lowering(ctx, indptr, indices, matdat, res, platform="cpu"):
     layout = default_layouts(ip_type.shape, idx_type.shape, md_type.shape)
     layout.insert(0, ())
     out = custom_call(
-        b"cpu_knn_f32",                                       # call target name
-        out_types=[res_type],
-        operands=[mlir.ir_constant(N), indptr, indices, matdat], 
-        operand_layouts=layout, 
-        result_layouts=default_layouts(res_type.shape),       # memory layout
-    )
+        b"cpu_knn_f32",  # call target name
+        result_types=[res_type],
+        operands=[mlir.ir_constant(N), indptr, indices, matdat],
+        operand_layouts=layout,
+        result_layouts=default_layouts(res_type.shape),  # memory layout
+    ).results
 
     return [out]
-
-
 
 
 # =======================================================
@@ -153,24 +160,27 @@ def _knn_v2_lowering(ctx, indptr, indices, matdat, res):
     # dispatch for float32 only
     if mat_dtype != jnp.float32:
         raise NotImplementedError(f"unsupported dtype: {mat_dtype}")
+    try:
+        import gpu_ops
 
-    if gpu_ops is None:
+        if gpu_ops is None:
+            raise ValueError("gpu_ops not compiled")
+
+        # create opaque descriptor for problem size
+        opaque = gpu_ops.build_knn_descriptor(N)
+
+        out = custom_call(
+            b"gpu_knn_v2_f32",  # call target name
+            out_types=[res_type],
+            operands=[indptr, indices, matdat],
+            operand_layouts=default_layouts(
+                ip_type.shape, idx_type.shape, md_type.shape
+            ),
+            result_layouts=default_layouts(res_type.shape),  # memory layout
+            backend_config=opaque,  # opaque descriptor
+        ).results
+    except:
         raise ValueError("gpu_ops not compiled")
-    
-    # create opaque descriptor for problem size
-    opaque = gpu_ops.build_knn_descriptor(N)
-
-    out = custom_call(
-        b"gpu_knn_v2_f32",                                   # call target name
-        out_types=[res_type],
-        operands=[indptr, indices, matdat], 
-        operand_layouts=default_layouts(ip_type.shape,
-                                         idx_type.shape, 
-                                         md_type.shape),
-        result_layouts=default_layouts(res_type.shape),   # memory layout
-        backend_config=opaque,                            # opaque descriptor
-    )
-
     # output must be iterable
     return [out]
 
@@ -179,46 +189,56 @@ def _knn_v2_lowering(ctx, indptr, indices, matdat, res):
 #             Registering KNN Primitive
 # =======================================================
 
+
+try:
+    import gpu_ops
+
+    is_gpu_available = True
+except ImportError:
+    is_gpu_available = False
+
 # register CPU XLA custom calls
 for _name, _value in cpu_ops.registrations().items():
     xla_client.register_custom_call_target(_name, _value, platform="cpu")
 
 # register GPU XLA custom calls
-for _name, _value in gpu_ops.registrations().items():
-    xla_client.register_custom_call_target(_name, _value, platform="gpu")
-
-
+if is_gpu_available:
+    for _name, _value in gpu_ops.registrations().items():
+        xla_client.register_custom_call_target(_name, _value, platform="gpu")
 # defining KNN primitive for JAX
 _knn_prim = core.Primitive("knn")
 _knn_prim.def_impl(partial(xla.apply_primitive, _knn_prim))
 _knn_prim.def_abstract_eval(_knn_abstract_eval)
 
 # connect XLA translation rules for JIT compilation
-mlir.register_lowering(_knn_prim, 
-                        partial(_knn_lowering, platform="gpu"),
-                        platform="gpu"
-)
-mlir.register_lowering(_knn_prim, 
-                        partial(_knn_lowering, platform="cpu"),
-                        platform="cpu"
+if is_gpu_available:
+    mlir.register_lowering(
+        _knn_prim, partial(_knn_lowering, platform="gpu"), platform="gpu"
+    )
+mlir.register_lowering(
+    _knn_prim, partial(_knn_lowering, platform="cpu"), platform="cpu"
 )
 
 
-# defining KNN v2 primitive
-_knn_v2_prim = core.Primitive("knn_v2")
-_knn_v2_prim.def_impl(partial(xla.apply_primitive, _knn_v2_prim))
-_knn_v2_prim.def_abstract_eval(_knn_v2_abstract_eval)
-mlir.register_lowering(_knn_v2_prim, _knn_v2_lowering, platform="gpu")
+if is_gpu_available:
+    # defining KNN v2 primitive
+    _knn_v2_prim = core.Primitive("knn_v2")
+    _knn_v2_prim.def_impl(partial(xla.apply_primitive, _knn_v2_prim))
+    _knn_v2_prim.def_abstract_eval(_knn_v2_abstract_eval)
+    mlir.register_lowering(_knn_v2_prim, _knn_v2_lowering, platform="gpu")
 
 
 # testing on a simple example
 # TODO: make proper tests for this
 if __name__ == "__main__":
-    foo = jnp.array([
-        [0.012, 0.3, 0.2],
-        [0.01, 0, 0.3],
-        [0.4, 0.1, 0.5],
-    ], dtype=jnp.float32)
+    foo = jnp.array(
+        [
+            [0.012, 0.3, 0.2],
+            [0.01, 0, 0.3],
+            [0.4, 0.1, 0.5],
+        ],
+        dtype=jnp.float32,
+    )
 
     foo = sparse.bcsr_fromdense(foo)
     x = knn(foo.indptr, foo.indices, foo.data, 3)

--- a/src/protax/classify.py
+++ b/src/protax/classify.py
@@ -1,5 +1,7 @@
+import os
+
 from .protax_utils import read_model_jax, read_query, read_baseline
-from .model import get_probs
+from .model import get_probs, get_log_probs
 import jax
 import jax.numpy as jnp
 import numpy as np
@@ -112,7 +114,7 @@ def compute_perplexity(qdir, model_dir, tax_dir, verbose=False):
             break  # EOF
         
         start = time.time()
-        probs = model.get_log_probs(q, ok, tree, params, segnum, N).block_until_ready()
+        probs = get_log_probs(q, ok, tree, params, segnum, N).block_until_ready()
         end = time.time()
 
 
@@ -142,6 +144,7 @@ if __name__ == "__main__":
     # tax_dir, model_dir, query_dir = protax_args[1:4]
 
     # testing for now
-    query_dir = r"/home/roy/Documents/PROTAX-dsets/30k_small/refs.aln"
+    
+    query_dir = r"FinPROTAX/FinPROTAX/modelCOIfull/refs.aln"
     classify_file(query_dir, "models/params/model.npz", "models/ref_db/taxonomy37k.npz") 
     compute_perplexity(query_dir, "models/params/model.npz", "models/ref_db/taxonomy37k.npz")

